datalink/setwatch.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH DATALINK 1 "December  1, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
setwatch, settime \- programs to send data to your Timex Datalink watch
.SH SYNOPSIS
.B setwatch
.RI [ WATCH_TYPE ]\ [ OPTIONS ]\ [ DATAFILE ]
.br
.B settime
.RI [ WATCH_TYPE ]
.SH DESCRIPTION
This manual page documents briefly the
.B setwatch
and
.B settime
commands.
This manual page was written for the Debian GNU/Linux distribution
because the original programs do not have manual pages.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBsetwatch\fP is the main program in the datalink suite that can send
data to your Timex Datalink watch.  It can send all types of data used
by these watches.
.PP
\fBsettime\fP is a simple utility that will just update the time on your
watch. It will only accept a watch-type argument (defaulting to
-ironman) and sends just the time.
.SH WATCH_TYPE
.TP
.B \-\-help
Show summary of options.
.TP
.B \-ironman
Use if you have an Ironman Datalink watch.  This is the default.
.TP
.B \-70, \-\-model70
Use if you have a model 70 watch.
.TP
.B \-150s, \-\-model150
Use if you have a model 150 watch.
.TP
.B \-150, \-\-model150s
Use if you have a model 150s watch.
.SH OPTIONS
Run "setwatch -help" for options summary.
.SH DATAFILE
By default, the datafile will be loaded from ~/.datalink/datafile.  Just
copy the example file (perhaps installed as /usr/doc/datalink/datafile?)
to this location and edit it as you please before running setwatch. 
Note that if you specify no options, nothing will happen.  Perhaps you
want to try the -all option to start.
.SH AUTHOR
The original datalink software was written by Karl R. Hakimian.
.PP
David Fries <dfries@mail.win.org> added support for the Ironman Datalink to
the datalink software and now is the maintainer.  http://datalink.fries.net/
is the current home.
.PP
This manual page was written by Steaphan Greene <stea@cs.binghamton.edu>,
for the Debian GNU/Linux system (but may be used by others).


datalink/changelog
07-21-2002:
	-Fixed the setwatch options of -app and +app to say appointments
	 instead of wristapp.
	-setwatch was trying to run svgablink with -150S instead of -150s
	-The -file (for creating a DEBUGOUTPUT) was broken, fixed.
07-17-2002: aka 1.0.0
	-Added header files for libc functions we are already using,
	 but weren't including the proper header files.
	-Switched to using tmpfile instead of tmpnam for the temporary
	 file, this will fix the compiler warnings.
	-The newer bison required a semicolon and would not compile
	 until it was added.
	-Replaced tmpnam() with mkstemp() because the compiler gave
	 warnings about tmpnam being a possible security hole.
	-Dependency information now goes in the file depend and will be
	 calculated if the file doesn't exist.
-- David Fries <dfries@mail.win.org>

12-16-2001:
	-Changed license to GPL (with permission from all 4 authors).
	-Added info to readme about GPL and about lack of support from Timex.
	-Created manpages for svgablink, serblink, setwatch, and settime
	 They still need work - I'll get to it when I can.
	-Changed settime default to ironman for consistency w/ setwatch.
	-Fixed apparent missing semicolon in dataread.y.
	-Added DESTDIR to Makefile to be more friendly with Debian packaging.
	-Split Makefile "install" rule into "install" (for basic datalink
	 user requirements) and "install-devel" for items needed only for
	 development.
	-Added Makefile "uninstall" rule.
-- Steaphan Greene <stea@cs.binghamton.edu>

02-14-2001:
The changes listed below from Frank Denis were incorporated with exception
of the time of the week information, I'll have to look into that later.
More important user visible changes are as follows,
	It will now only search in the users path for serblink and svgablink
		not any hard-coded paths.  Previous behavior was to try
		the user path and if it wasn't found, try executing
		the program in the current working directory.
	settime is now installed by default and will accept a command line for
		the different types of watches
	For me, svgablink no longer hangs on exit
-- David Fries <dfries@mail.win.org>

02-04-2001:
I don't see the problem with the day of week.  I'm reverting back to
the previous code.  Will have to investigate sometime.  On my watch
the code is correct as it used to be (and is now again).
-- David Fries

04-22-2000:
Fixed a bug with the day of the week that was shifted by one. Please report
if it broke something.
-- Frank Denis

04-20-2000:
Fixed the path to svgablink.
The datalink 150 S protocol was implemented but the "-150s" option was
rejected in setwatch. Fixed.
-- Frank Denis

04-17-2000:
Add support for Datalink 150 S watches
-- Frank Denis


04-14-2000:
Lots, lots, lots of cleanups to the source code in order to make it compile
on modern compilers.
-- Frank Denis


01-05-2000:
Added a patch from Scottie Shore <sshore@escape.ca> for a Y2K problem.
As localtime(3) reads,
       tm_year
              The number of years since 1900.
The watch is expecting the last two digits and displays A0 for the year instead of 00.  Fix is to take the year % 100.
-- David Fries


12-11-1999:
Added a patch from Danilo Fiorenzano <danilo@telusplanet.net> to
Added new function "maxPriority()" to switch to SCHED_RR kernel scheduling
policy at maximum priority and lock all pages into real memory, in order
to improve timing accuracy.
-- David Fries


datalink/setwatch.c
/* setwatch: Example of Linux SDK for Timex watch.             */
/* Written by Karl R. Hakimian 10/3/96                         */
/*                                                             */
/* Now warranty expressed or implied, use it as you see fit.   */

#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/time.h>
#include <unistd.h>
#include <datalink.h>

my_error_proc(msg)
char *msg;
{
	fprintf(stderr, "ERROR: %s\n", msg);
	exit(-1);
}

my_warn_proc(msg)
char *msg;
{
	fprintf(stderr, "WARNING: %s\n", msg);
	return(0);
}

ListPtr
set_time(wi, times)
WatchInfoPtr wi;
ListPtr times;
{
	time_t now;
	time_t tztime;
	struct tm *time_s;
	struct timezone tz;
	ItemPtr tp;
	int isdst;
	char buf[1024];
	int i;

	now = time(NULL) + wi->time_adjust; /* Offset for sending to watch. */
	time_s = localtime(&now);
	isdst = (time_s->tm_isdst) ? 1 : 0;

	if (!times->count) {
		tp = dl_new_item(wi, DL_TIME_TYPE);
		tp->data.time.label = tzname[isdst];
		tp->data.time.offset = timezone/60;
		tp->data.time.offset = (isdst) ? (timezone/60) - 60 : (timezone/60);
		tp->data.time.hour_fmt = 1;
		dl_add_to_list(times, tp);
		tp = dl_new_item(wi, DL_TIME_TYPE);
		tp->data.time.label = tzname[1 - isdst];
		tp->data.time.offset = (isdst) ? (timezone/60) : (timezone/60) - 60;
		tp->data.time.hour_fmt = 1;
		dl_add_to_list(times, tp);
	}

	for (i = 0, tp = times->first; i < times->count; i++, tp = tp->next) {

		if (!*tp->data.time.label) {
			tp->data.time.label = tzname[isdst];
			tp->data.time.offset = timezone/60;
		}

		tztime = now - tp->data.time.offset*60;
		time_s = gmtime(&tztime);
		tp->data.time.hours = time_s->tm_hour;
		tp->data.time.minutes = time_s->tm_min;
		tp->data.time.seconds = time_s->tm_sec;
		tp->data.time.month = time_s->tm_mon + 1;
		tp->data.time.day = time_s->tm_mday;
		tp->data.time.year = time_s->tm_year;
		tp->data.time.dow = (time_s->tm_wday + 6)%7;
		tp->data.time.download = 1;
	}

	return(times);
}

#define TIME 0x001
#define ALARM 0x002
#define APP 0x004
#define TODO 0x008
#define PHONE 0x010
#define ANNIV 0x020
#define SYSTEM 0x040
#define WRISTAPP 0x080
#define MELODY 0x100
#define TIMER 0x200

#define ALL 0x1FF
#define DB 0x03C
/*
#define DEFAULT 0x03F
*/
#define DEFAULT 0

main(argc, argv)
int argc;
char **argv;
{
	char *prog = argv[0];
	WatchInfoPtr wi;
	ListPtr times, alarms, timers, apps, todos, phones, annivs;
	ListPtr system, wristapp, melody;
	char datafile[1024];
	int output = SVGA_BLINK;
	int flags = DEFAULT;
	int (*app_sort)() = dl_app_by_datetime;
	int (*todo_sort)() = dl_todo_by_prio;
	int (*phone_sort)() = dl_phone_by_label;
	int (*anniv_sort)() = dl_anniv_by_date;
	int type = DATALINK_IRONMAN;

	while (argc > 1 && (*argv[1] == '-' || *argv[1] == '+')) {

		if (strcmp("-all", argv[1]) == 0)
			flags = ALL;
		else if (strcmp("-db", argv[1]) == 0)
			flags = DB;
		else if (strcmp("+db", argv[1]) == 0)
			flags |= DB;
		else if (strcmp("-time", argv[1]) == 0)
			flags = TIME;
		else if (strcmp("+time", argv[1]) == 0)
			flags |= TIME;
		else if (strcmp("-alarm", argv[1]) == 0)
			flags = ALARM;
		else if (strcmp("+alarm", argv[1]) == 0)
			flags |= ALARM;
		else if (strcmp("-timer", argv[1]) == 0)
			flags = TIMER;
		else if (strcmp("+timer", argv[1]) == 0)
			flags |= TIMER;
		else if (strcmp("-wristapp", argv[1]) == 0)
			flags = WRISTAPP;
		else if (strcmp("+wristapp", argv[1]) == 0)
			flags |= WRISTAPP;
		else if (strcmp("-app", argv[1]) == 0)
			flags = WRISTAPP;
		else if (strcmp("+app", argv[1]) == 0)
			flags |= WRISTAPP;
		else if (strcmp("-melody", argv[1]) == 0)
			flags = MELODY;
		else if (strcmp("+melody", argv[1]) == 0)
			flags |= MELODY;
		else if (strcmp("-system", argv[1]) == 0)
			flags = SYSTEM;
		else if (strcmp("+system", argv[1]) == 0)
			flags |= SYSTEM;
		else if (strcmp("-reset", argv[1]) == 0) {
			flags = 0;
			app_sort = todo_sort = phone_sort = anniv_sort = NULL;
		}
		else if (strcmp("-sort-app-by-datetime", argv[1]) == 0)
			app_sort = dl_app_by_datetime;
		else if (strcmp("-sort-app-by-label", argv[1]) == 0)
			app_sort = dl_app_by_label;
		else if (strcmp("-sort-todo-by-label", argv[1]) == 0)
			todo_sort = dl_todo_by_label;
		else if (strcmp("-sort-todo-by-prio", argv[1]) == 0)
			todo_sort = dl_todo_by_prio;
		else if (strcmp("-sort-phone", argv[1]) == 0)
			phone_sort = dl_phone_by_label;
		else if (strcmp("-sort-anniv-by-date", argv[1]) == 0)
			anniv_sort = dl_anniv_by_date;
		else if (strcmp("-sort-anniv-by-label", argv[1]) == 0)
			anniv_sort = dl_anniv_by_label;
		else if (strcmp("-file", argv[1]) == 0)
			output = BLINK_FILE;
		else if (strcmp("-serial", argv[1]) == 0)
			output = SER_BLINK;
		else if (strcmp("-model70", argv[1]) == 0)
			type = DATALINK_70;
		else if (strcmp("-70", argv[1]) == 0)
			type = DATALINK_70;
		else if (strcmp("-ironman", argv[1]) == 0)
			type = DATALINK_IRONMAN;
		else {
			fprintf(stderr, "%s: Unknown option (%s).\n", prog, argv[1]);
			exit(-1);
		}

		argc--;
		argv++;
	}

	if (argc == 2)
		strcpy(datafile, argv[1]);
	else if (argc > 2) {
		fprintf(stderr, "Usage: %s [datafile]\n", argv[0]);
		exit(-1);
	}
	else {
		strcpy(datafile, getenv("HOME"));
		strcat(datafile, "/");
		strcat(datafile, ".datalink/datafile");
	}
/*
 Use my error proc to exit on error and my warn proc to ignore warnings
*/
	dl_set_error(my_error_proc);
	dl_set_warn(my_warn_proc);

	wi = dl_read_save(datafile, type, &times, &alarms, &timers, &apps, &todos, &phones,
		&annivs, &system, &wristapp, &melody);

/* Mark for download. */

	if ((flags&ALARM) && alarms->count)
		alarms->download = 1;

	if ((flags&TIMER) && timers->count)
		timers->download = 1;

	if ((flags&APP) && apps->count) {
		apps->download = 1;

		if (app_sort)
			dl_sort(apps, app_sort);

	}

	if ((flags&TODO) && todos->count) {
		todos->download = 1;

		if (todo_sort)
			dl_sort(todos, todo_sort);

	}

	if ((flags&PHONE) && phones->count) {
		phones->download = 1;

		if (phone_sort)
			dl_sort(phones, phone_sort);

	}

	if ((flags&ANNIV) && annivs->count) {
		annivs->download = 1;

		if (anniv_sort)
			dl_sort(annivs, anniv_sort);

	}

	if ((flags&SYSTEM) && system->count)
		system->download = 1;

	if ((flags&WRISTAPP) && wristapp->count)
		wristapp->download = 1;

	if ((flags&MELODY) && melody->count)
		melody->download = 1;

	if ((flags&TIME)) {
		times = set_time(wi, times);
		times->download = 1;
	}
	else
		times = NULL;
/*
 Send it to the watch
*/
	dl_init_download(wi, times, alarms, timers, apps, todos, phones, annivs, system,
		wristapp, melody);
	dl_send_data(wi, output);
	exit(0);
}
